lib/ipam/ipam_block_reader_writer.go

```diff
@@ -347,6 +345,76 @@ func (rw blockReaderWriter) releaseBlockAffinity(ctx context.Context, host strin
 	return nil
 }
 
+func (rw blockReaderWriter) queryAffinity(ctx context.Context, k model.Key, revision string) (*model.KVPair, error) {
+	aff, err := rw.client.Get(ctx, k, revision)
+	if err != nil {
+		return nil, err
+	}
+
+	if aff.Value.(*model.BlockAffinity).State != model.StateDeleted {
+		// We're all good - return the afifnity.
+		return aff, nil
+	}
+
+	// If the affinity is marked as deleted, we need to delete it.
+	if err := rw.deleteAffinity(ctx, aff); err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+			// Already deleted - we're good.
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	// Should return a "does not exist" error.
+	return rw.client.Get(ctx, k, revision)
+}
+
+func (rw blockReaderWriter) updateAffinity(ctx context.Context, aff *model.KVPair) (*model.KVPair, error) {
+	return rw.client.Update(ctx, aff)
+}
+
+func (rw blockReaderWriter) deleteAffinity(ctx context.Context, aff *model.KVPair) error {
+	// Mark the affinity as block deleted.
+	aff.Value.(*model.BlockAffinity).State = model.StateDeleted
+	aff, err := rw.client.Update(ctx, aff)
+	if err != nil {
+		return err
+	}
+
+	// We've removed / updated the block, so perform a compare-and-delete on the BlockAffinity.
+	_, err = rw.client.Delete(ctx, aff.Key, aff.Revision)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (rw blockReaderWriter) updateBlock(ctx context.Context, b *model.KVPair) (*model.KVPair, error) {
+	if b.Value.(*model.AllocationBlock).Deleting {
+		// If the block is in the process of being deleted, we should return
+		// a conflict error and retry.
+		return nil, cerrors.ErrorResourceUpdateConflict{Err: fmt.Errorf("Resource being deleted"), Identifier: b.Key}
+	}
+
+	return rw.client.Update(ctx, b)
+}
+
+func (rw blockReaderWriter) deleteBlock(ctx context.Context, b *model.KVPair) error {
```

Comment:
We need to implement a `queryBlock` here as well (similar to query affinity above). 

And, we need to do all of these for IPAM handle as well.